src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -39,195 +50,285 @@
+    }
+
+    private @CheckForNull <T extends SCMSourceTrait> T getTrait(@Nonnull Class<T> traitClass) {
+        for (SCMSourceTrait t: traits) {
+            if (traitClass.isInstance(t)) {
+                return traitClass.cast(t);
+            }
         }
-        final HgExe hg = new HgExe(inst, credentials, launcher, node, listener, new EnvVars());
-        try {
-            String heads = hg.popen(cache, listener, true, new ArgumentListBuilder("heads", "--template", "{node} {branch}\\n"));
-            Pattern p = Pattern.compile(Util.fixNull(branchPattern).length() == 0 ? ".+" : branchPattern);
-            for (String line : heads.split("\r?\n")) {
-                final String[] nodeBranch = line.split(" ", 2);
-                final String name = nodeBranch[1];
-                if (p.matcher(name).matches()) {
-                    listener.getLogger().println("Found branch " + name);
-                    SCMHead branch = new SCMHead(name);
-                    if (criteria != null) {
-                        SCMProbe probe = new SCMProbe() {
-                            @NonNull
-                            @Override
-                            public SCMProbeStat stat(@NonNull String path) throws IOException {
-                                try {
-                                    String files = hg.popen(cache, listener, true,
-                                            new ArgumentListBuilder("locate", "-r", nodeBranch[0], "-I", "path:" + path));
-                                    if (StringUtils.isBlank(files)) {
-                                        return SCMProbeStat.fromType(SCMFile.Type.NONEXISTENT);
+        return null;
+    }
+
+    @Override protected void retrieve(@CheckForNull SCMSourceCriteria criteria, @Nonnull SCMHeadObserver observer,
+                            @CheckForNull SCMHeadEvent<?> event, @Nonnull final TaskListener listener)
+            throws IOException, InterruptedException {
+        try (MercurialSCMSourceRequest request= new MercurialSCMSourceContext<>(criteria, observer)
+                .withTraits(traits)
+                .newRequest(this, listener) ) {
+            MercurialInstallation inst = MercurialSCM.findInstallation(request.installation());
+            if (inst == null) {
+                listener.error("No configured Mercurial installation");
+                return;
+            }
+            if (!inst.isUseCaches()) {
+                listener.error("Mercurial installation " + request.installation() + " does not support caches");
+                return;
+            }
+            final Node node = Jenkins.getInstance();
+            if (node == null) { // Should not happen BTW
+                listener.error("Cannot retrieve the Jenkins master node");
+                return;
+            }
+            Launcher launcher = node.createLauncher(listener);
+            StandardUsernameCredentials credentials = getCredentials(request.credentialsId());
```

Comment:
I think this regressed credentials handling: #108. Means that authenticated repository access is not being tested.